designs/set-design.md

```diff
@@ -0,0 +1,195 @@
+# Set Dict Tuple Design
+## Change to AST and Ir
+We made changes to ast.ts and ir.ts to include set, dictionary and tuples. Though we will focus on set but we have included all three types in case other group might refer to these types. Changes in ast and ir are the same as we have not included optimization currently.
+
+Similar to class implementation in PA3, we add three new types under the Type we declared as:
+```
+export type Type =
+    ……
+  | { tag: "set"; content_type: Type }
+  | { tag: "dict"; key: Type; value: Type }
+  | { tag: "tuple"; contentTypes: Array<Type> }
+```
+It can be used in typechecker to decide the type for each inirtialized objects. And now we have made a constraint that all elements in the set or dict or tuple should be of the same type. We might allow various types in the future.
+
+Then in the expression, the change we have made is as:
+```
+export type Expr<A> =
+    ……
+ | { a?: A, tag: "set_expr", contents: Array<Expr<A>> }
+ | { a?: A, tag: "tuple_expr", contents: Array<Expr<A>> }
+ | { a?: A, tag: "dict_expr", entries: Array<[Expr<A>, Expr<A>]> }
+```
+## Value Representation and Memory Management
+All Values are stored as 8-byte i32. Memory management should use the interface provided by the memory group, the memory usage should be kept below 30% to maintain the efficiency of the hash function. 
```

Comment:
I feel like this is the only mention of hashing/a hash table. How is that going to work?